### DISCUSSION TITLE
Create array of literals from zod objects

### DISCUSSION THREAD

**Discussion by theetrain** [colinhacks/zod#2283] [Q&A]
Hello, my use case is to define a zod schema to parse form fields after they're submitted. Some form fields represent fields in a deep object to be added to a JSON column in MySQL. Consider this schema:

```js
import { z } from 'zod';

const price = z.number().multipleOf(0.01).nonnegative().nullable();

const productSchema = z.object({
  Name: z.string(),
  Order: z.number().int().nullable(),
  Price: z
    .object({
      cad: price,
      usd: price,
      eur: price,
    })
    .partial(),
  Details: z.record(z.string()).nullable(),
});
```

I have form elements set up (oversimplified example):

```html
<input name="Name" />
<input name="Order" />
<input name="Price[cad]" />
<input name="Price[usd]" />
<!-- ... -->
```

In my back-end I convert `Price[cad]` to `Price: { cad: value }`. I would like to have type safety on these `name` values, so I attempted to create a flattened array:

```js
const productSchemaFlattened = Object.entries(productSchema.shape)
  .map((el) => {
    if (el[1]?.shape) {
      return Object.keys(el[1].shape).map((item) => `${el[0]}[${item}]`);
    }
    return el[0];
  })
  .flat();
```

And although `productSchemaFlattened` does produce the array of literal values I want (`['Name', 'Order', 'Price[cad]', ...]`), I don't have the appropriate autocompletion or type safety I was hoping for—it displays zod methods instead of string literals. How do I produce this flattened array to be reusable as a type?

I'm new to zod, so my approach here may be wrong as well. Whether I have deep object keys, or a schema field with `z.object().passthrough()`, I would like to take the known keys and turn that into a reusable array of literals for easier form maintenance. 

Demo: https://stackblitz.com/edit/node-ykpkqd?file=index.js
Somewhat related: https://github.com/colinhacks/zod/discussions/839

**Comment by JacobWeisenburger** [comment]
perhaps this would help: https://github.com/colinhacks/zod/discussions/2134

**Reply by theetrain** [reply]
Thanks for your reply. Similarly to my example, the linked utility will flatten the deep-nested keys into an array of strings, but the typings are not there (also mentioned in https://github.com/colinhacks/zod/discussions/2134#discussioncomment-5199845)

This could be a TypeScript shortcoming; maybe it cannot infer string literals that are generated 🤔 such as from statements like this:

```js
const nested = value instanceof z.ZodType ? zodKeys(value).map(subKey => `${key}.${subKey}`) : [];
                                                                        // ^^^^^^^^^^^^^^^
```

Here are the autocomplete suggestions I see:

<img width="776" alt="image" src="https://user-images.githubusercontent.com/12798751/229213402-d6015d43-ba19-4b3a-9f4b-9c799cddf1ad.png">

**Comment by JacobWeisenburger** [comment]
This seems to be a limitation of TypeScript.

```ts
const obj = {
    fooBase: 42,
    fooNest: {
        bar: {
            baz: 42
        }
    }
} as const

type GetKeys<Obj extends Record<string,unknown>> = {
    [ key in keyof Obj ]: Obj[ key ] extends object
    ? `${ key }.${ GetKeys<Obj[ key ]> }` // ts error
    : Obj[ key ]
}
type Keys = GetKeys<typeof obj>
```

More info: https://stackoverflow.com/a/55128956/5321863

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
